aws-api/src/main/java/com/amplifyframework/api/aws/GsonListDeserializer.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package com.amplifyframework.api.aws;
+
+import com.google.gson.Gson;
+import com.google.gson.JsonArray;
+import com.google.gson.JsonDeserializationContext;
+import com.google.gson.JsonDeserializer;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParseException;
+
+import java.lang.reflect.ParameterizedType;
+import java.lang.reflect.Type;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * Custom list deserializer since some lists come back not as an array of the items but as an object which contains
+ * an items property with the list of items and a nextToken property for pagination purposes.
+ */
+final class GsonListDeserializer implements JsonDeserializer<List<Object>> {
```

Comment:
Strangely I could have sworn this deserializer was not actually used, and instead the logic for deserializing the "items" array was handled [in GsonGraphQLResponseFactory](https://github.com/aws-amplify/amplify-android/blob/4d85b459af2539c6d6b0586228fd7ed6cc320986/aws-api/src/main/java/com/amplifyframework/api/aws/GsonGraphQLResponseFactory.java#L141-L142). 

I'm guessing there's a use case that I'm not familiar with yet - I'll take a closer look at the integration tests today to try and understand this further.

Reply:
I was surprised by this, too. I was pretty sure we have some unit tests that contain listed items. Not sure where the discrepancy is coming from!

Reply:
Ok, I figured this out. 

[`buildSingleArrayResponse`](https://github.com/aws-amplify/amplify-android/blob/be8560576ed6d22da73b5b7239e942401309c7be/aws-api/src/main/java/com/amplifyframework/api/aws/GsonGraphQLResponseFactory.java#L95-L116) automatically skips the "items" level before parsing it with Gson.

[`buildSingleItemResponse`](https://github.com/aws-amplify/amplify-android/blob/be8560576ed6d22da73b5b7239e942401309c7be/aws-api/src/main/java/com/amplifyframework/api/aws/GsonGraphQLResponseFactory.java#L57-L93) does not, so it relies on `GsonListDeserializer`.

I guess that means we have no unit tests for `buildSingleItemResponse`, so definitely something to add at some point.